server/src/main/java/org/elasticsearch/rest/RestController.java

```diff
@@ -255,11 +255,19 @@ private void sendContentTypeErrorMessage(@Nullable List<String> contentTypeHeade
     }
 
     private void tryAllHandlers(final RestRequest request, final RestChannel channel, final ThreadContext threadContext) throws Exception {
-        for (String key : headersToCopy) {
-            String httpHeader = request.header(key);
-            if (httpHeader != null) {
-                threadContext.putHeader(key, httpHeader);
+        try {
+            for (RestHeaderDefinition restHeader : headersToCopy) {
+                final String name = restHeader.getName();
+                String httpHeader = restHeader.isMultiValueAllowed() ? request.getAllHeaderValuesAsString(name) :
+                    request.getSingleValuedHeader(name);
+                if (httpHeader != null) {
+                    threadContext.putHeader(name, httpHeader);
+                }
             }
+        } catch (IllegalStateException e) {
```

Comment:
It's too bad we have to use an unchecked exception for flow control here. Can we look for another way? For example, one suggestion would be to pull the multi-valued validation into this method instead of holding it in RestRequest#getSingleValuedHeader.

Or, if this is not palatable, or if we can't find another way, can `RestRequest#getSingleValuedHeader` at least declare this exception in its Javadocs?